cassandra/src/test/java/org/apache/calcite/test/CassandraExtension.java

```diff
@@ -0,0 +1,218 @@
+  }
+
+  private static class CassandraResource implements ExtensionContext.Store.CloseableResource {
+    private final Session session;
+    private final Cluster cluster;
+
+    private CassandraResource() {
+      startCassandra();
+      this.cluster = EmbeddedCassandraServerHelper.getCluster();
+      this.session = EmbeddedCassandraServerHelper.getSession();
+    }
+
+    /**
+     * Best effort to gracefully shutdown <strong>embedded</strong> cassandra cluster.
+     *
+     * Since it uses many static variables as well as {@link System#exit(int)} during close,
+     * clean shutdown (as part of unit test) is not straightforward.
+     */
+    @Override public void close() {
+
+      session.close();
+      cluster.close();
+
+      CassandraDaemon daemon = extractDaemon();
+      if (daemon.thriftServer != null) {
+        daemon.thriftServer.stop();
+      }
+      daemon.stopNativeTransport();
+
+      StorageService storage = StorageService.instance;
+      storage.setRpcReady(false);
+      storage.stopClient();
+      storage.stopTransports();
+      try {
+        storage.drain(); // try to close all resources
+      } catch (IOException | ExecutionException e) {
+        throw new RuntimeException(e);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+      StageManager.shutdownNow();
+
+      if (FBUtilities.isWindows) {
+        // for some reason .toDelete stale folder is not deleted on cassandra shutdown
+        // doing it manually here
+        WindowsFailedSnapshotTracker.resetForTests();
+        WindowsFailedSnapshotTracker.deleteOldSnapshots();
+      }
+    }
+
+    private static void startCassandra() {
+      // This static init is necessary otherwise tests fail with CassandraUnit in IntelliJ (jdk10)
+      // should be called right after constructor
+      // NullPointerException for DatabaseDescriptor.getDiskFailurePolicy
+      // for more info see
+      // https://github.com/jsevellec/cassandra-unit/issues/249
+      // https://github.com/jsevellec/cassandra-unit/issues/221
+      DatabaseDescriptor.daemonInitialization();
+
+      // Apache Jenkins often fails with
+      // Cassandra daemon did not start within timeout (20 sec by default)
+      try {
+        EmbeddedCassandraServerHelper.startEmbeddedCassandra(Duration.ofMinutes(1).toMillis());
+      } catch (TTransportException e) {
+        throw new RuntimeException(e);
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+    }
+
+    /**
+     * Extract {@link CassandraDaemon} instance using reflection. It will be used
+     * to shutdown the cluster
+     */
+    private static CassandraDaemon extractDaemon() {
+      try {
+        Field field = EmbeddedCassandraServerHelper.class.getDeclaredField("cassandraDaemon");
+        field.setAccessible(true);
+        CassandraDaemon daemon = (CassandraDaemon) field.get(null);
```

Comment:
Is this really required?
I thought it should have a way to properly shutdown cassandra

Reply:
Well I'm not sure Cassandra was ever meant to be embedded ([example](https://github.com/apache/cassandra/blob/cassandra-3.11/src/java/org/apache/cassandra/service/CassandraDaemon.java#L577) from CassandraDaemon.java):

```java
// code from CassandraDaemon.java

public void stop() {
// ... 
// On windows, we need to stop the entire system as prunsrv doesn't have the jsvc hooks
// We rely on the shutdown hook to drain the node
  if (FBUtilities.isWindows)
         System.exit(0);
// .... 
}
```

`System.exit()` inside `Service.stop/close` is very telling. So one has to find work-arounds to correctly shutdown cassandra without stopping JVM

Reply:
We use [cassandra-unit](https://github.com/jsevellec/cassandra-unit) to manage cassandra unfortunately they don't expose `CassandraDaemon`.

Perhaps one can start Cassandra manually without cassandra-unit but I didn't want to introduce too many changes.